### ISSUE TITLE
traverse cannot be second in pipe

### ISSUE THREAD

**Issue by pickatag** [plataformatec/nimble_parsec#29]
The following code works OK.
```
defmodule MyParser do
  @moduledoc false

  @debug false

  import NimbleParsec

  white_space =
    ignore(ascii_string([9, 32], min: 1))

  language_code =
    optional(white_space) |>      # Needs to be here in order to make traverse work
    ascii_string([?a..?z], 2) |>
    traverse(:check_language_code) |>
    tag(:language)

  territory_code =
    ascii_string([?A..?Z], 2) |>
    tag(:territory)

  locale =
    choice([language_code, territory_code]) |>
    ignore(string(":"))

  locale_text =
    optional(locale) |>
    optional(white_space) |>
    utf8_string([], min: 1)

  defparsec :locale_text, locale_text, debug: @debug

  defp check_language_code(_rest, [language_code], context, _line, _offset) do
    case Cldr.validate_locale(language_code) do
      {:ok, %Cldr.LanguageTag{cldr_locale_name: cldr_locale_name}} -> {cldr_locale_name |> String.to_atom |> List.wrap, context}
      {:error, _} -> {:error, "Not a valid language code (#{language_code})"}
    end
  end
end
```

`MyParser.locale_text("en:A text")` works and returns a valid language code and `MyParser.locale_text("qq:A text")` fails with an error.

But in order to make it work, I had to put an extra part in front of `ascii_string([?a..?z], 2)` in ` language_code` above, otherwise the compiler failed.

I have no idea of the difference and for me it does not matter, but it is not clear to me why `traverse` fails being second in the pipe chain.

**Comment by josevalim** [comment]
@pickatag just to avoid any confusion, can you also post the code that failed? I assume the code above is the one working. Thank you!

**Comment by pickatag** [comment]
Yes, of course, sorry. This code fails in compilation. (I removed dependency to ex_Cldr as well)
```
defmodule MyParser do
  @moduledoc false

  @debug false

  import NimbleParsec

  white_space =
    ignore(ascii_string([9, 32], min: 1))

  language_code =
    ascii_string([?a..?z], 2) |>
    traverse(:check_language_code) |>
    tag(:language)

  territory_code =
    ascii_string([?A..?Z], 2) |>
    tag(:territory)

  locale =
    choice([language_code, territory_code]) |>
    ignore(string(":"))

  locale_text =
    optional(locale) |>
    optional(white_space) |>
    utf8_string([], min: 1)

  defparsec :locale_text, locale_text, debug: @debug

  defp check_language_code(_rest, [language_code], context, _line, _offset) do
    {language_code |> String.to_atom |> List.wrap, context}
  end
end
```

**Comment by josevalim** [comment]
Thanks for the example! It has been fixed and 0.3.2 has been released!
